### ISSUE TITLE
Incorrect function parameter not detected

### ISSUE THREAD

**Issue by TeamworkGuy2** [Microsoft/TypeScript#13248]
**TypeScript Version:**  (maybe 2.1.x - typescriptlang.org playground version as of 2017-1-2)

**Code**
[typescriptlang.org playground](http://www.typescriptlang.org/play/index.html#src=%5B1%2C%202%2C%203%5D.map(mapper)%3B%20%2F%2F%20fine%3F%3F%3F%0D%0A%5B1%2C%202%2C%203%5D.map((obj%2C%20idx)%20%3D%3E%20mapper(obj%2C%20idx))%3B%20%2F%2F%20%3C%3D%20error%0D%0A%0D%0Atype%20Tag%20%3D%20number%20%26%20%7B%20_tagBrand%3A%20any%3B%20%7D%3B%0D%0A%0D%0Afunction%20mapper%3CE%3E(obj%3A%20E%2C%20type%3A%20Tag)%20%7B%0D%0A%20%20%20%20return%20obj%3B%0D%0A%7D)
```ts
[1, 2, 3].map(mapper); // fine???
[1, 2, 3].map((obj, idx) => mapper(obj, idx)); // <= error

type Tag = number & { _tagBrand: any; };

function mapper<E>(obj: E, type: Tag) {
    return obj;
}
```

**Expected behavior:**
The first `map()` call on line 1 should throw the same compile error as the second call, since the second parameter of the mapper function does not match the second parameter of the function expected by Array.map()

**Actual behavior:**
No compile error on the first map() call when it should.

Found while working on nominal types based on [basarat's git book 'Nominal Typings' section](https://basarat.gitbooks.io/typescript/content/docs/tips/nominalTyping.html). 

Reproduced in VS 2015 update 3 TypeScript 2.1.4 and the typescriptlang.org playground.

**Comment by akarzazi** [comment]
Function parameters are Bivariants
Here is a sample 

```ts
let fct: (x: number) => void = (a: Tag) => console.log(a); // ok 
let fft: (x: Tag) => void = (a: number) => console.log(a); // ok also
```

Read more :
http://www.typescriptlang.org/docs/handbook/type-compatibility.html
-> Function Parameter Bivariance

**Comment by aleksey-bykov** [comment]
> function parameters are bivariants

and this is a shame

**Comment by TeamworkGuy2** [comment]
Thanks @akarzazi, it definitely seems like a strange design, but TypeScript is pretty awesome so I'm not complaining :)
